# DS-SF-34 | 02 | The `pandas` Library | Codealong | Starter Code

(http://pandas.pydata.org/pandas-docs/stable)

## Part A - Introduction to `pandas`

In [1]:
import numpy as np
import pandas as pd
import re

> ## `pd.read_csv()`: load datasets from files (or even over the Internet)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)

In [2]:
import pandas as pd
df = pd.read_csv('../datasets/dataset-02-zillow-starter.csv')

> ## `DataFrame`

Let's check `df`'s type:

In [3]:

# TODO
type(df)

pandas.core.frame.DataFrame

`df` is a `DataFrame`.  (http://pandas.pydata.org/pandas-docs/stable/dsintro.html)

A `DataFrame` stores tabular data.  Let's have a look at its content:

In [8]:
df

ID                                       Address  Latitude  \
0      15063471     55 Vandewater St APT 9, San Francisco, CA  37805103   
1      15063505           740 Francisco St, San Francisco, CA  37804420   
2      15063609           819 Francisco St, San Francisco, CA  37803728   
3      15064044      199 Chestnut St APT 5, San Francisco, CA  37804392   
4      15064257    111 Chestnut St APT 403, San Francisco, CA  37804240   
5      15064295    111 Chestnut St APT 702, San Francisco, CA  37804240   
6      15064391     1821 Grant Ave APT 101, San Francisco, CA  37803748   
7      15064536        2300 Leavenworth St, San Francisco, CA  37802408   
8      15064640       1047-1049 Lombard St, San Francisco, CA  37801889   
9      15064669        1055 Lombard St # C, San Francisco, CA  37801873   
10     15065032    101 Lombard St APT 116E, San Francisco, CA  37803470   
11     15065140   1407 Montgomery St APT 2, San Francisco, CA  37802299   
12     15065727   1360 Montgomery St APT 1, San Francisco, CA  37801802   
13     15065810        1327-1329 Kearny St, San Francisco, CA  37800260   
14     15065946     44 Macondray Ln APT 2E, San Francisco, CA  37799474   
15     15066001      1070 Green St APT 201, San Francisco, CA  37798792   
16     15066036     1070 Green St APT 1402, San Francisco, CA  37798792   
17     15066085  1907 Leavenworth St APT 8, San Francisco, CA  37798645   
18     15066201        1101 Green St # 303, San Francisco, CA  37798234   
19     15066281              1807 Jones St, San Francisco, CA  37797920   
20     15066344       999 Green St APT 905, San Francisco, CA  37798451   
21     15066608     455 Vallejo St APT 111, San Francisco, CA  37798796   
22     15066839       1163-1165 Vallejo St, San Francisco, CA  37797445   
23     15066959           1300 Pacific Ave, San Francisco, CA  37795805   
24     15067302             948 Jackson St, San Francisco, CA  37795662   
25     15067583     946 Stockton St APT 6B, San Francisco, CA  37794715   
26     15067604     946 Stockton St APT 9E, San Francisco, CA  37794715   
27     15067755       1250 Clay St APT 306, San Francisco, CA  37793575   
28     15067890      1250 Jones St APT 503, San Francisco, CA  37793071   
29     15067896      1250 Jones St APT 802, San Francisco, CA  37793071   
..          ...                                           ...       ...   
970  2100992607               2137 Pine St, San Francisco, CA  37788242   
971  2100994004              409 Miguel St, San Francisco, CA  37736625   
972  2101028916            251 Missouri St, San Francisco, CA  37763202   
973  2101094207            1650 Florida St, San Francisco, CA  37746898   
974  2101105851          41 Collingwood St, San Francisco, CA  37761579   
975  2101116542             2753 Bryant St, San Francisco, CA  37750270   
976  2101147349            244 Precita Ave, San Francisco, CA  37747330   
977  2101153805                6 Rossi Ave, San Francisco, CA  37779494   
978  2101156735            333 Potrero Ave, San Francisco, CA  37765266   
979  2101167573              1340 Union St, San Francisco, CA  37798930   
980  2101169168            688 Douglass St, San Francisco, CA  37752803   
981  2101175813              414 Linden St, San Francisco, CA  37776218   
982  2101331127      1408 31st Ave APT 303, San Francisco, CA  37760994   
983  2101505254               333 Beale St, San Francisco, CA  37788280   
984  2101588805             82 Alvarado St, San Francisco, CA  37754504   
985  2101807538         2619 Sacramento St, San Francisco, CA  37789371   
986  2102385068            129 Central Ave, San Francisco, CA  37770870   
987  2102906272              444 Linden St, San Francisco, CA  37776170   
988  2106229715              2595 38th Ave, San Francisco, CA  37738600   
989  2108411930               3500 19th St, San Francisco, CA  37760100   
990  2108546010      1738 Lombard St APT 4, San Francisco, CA  37800910   
991  2114498518    201 Sansome St UNIT

> ## `.head()`: first 5 (default) rows

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.head.html)

In [9]:
df.head()

ID                                     Address  Latitude  Longitude  \
0  15063471   55 Vandewater St APT 9, San Francisco, CA  37805103 -122412856   
1  15063505         740 Francisco St, San Francisco, CA  37804420 -122417389   
2  15063609         819 Francisco St, San Francisco, CA  37803728 -122419055   
3  15064044    199 Chestnut St APT 5, San Francisco, CA  37804392 -122406590   
4  15064257  111 Chestnut St APT 403, San Francisco, CA  37804240 -122405509   

  DateOfSale  SalePrice SalePriceUnit IsAStudio  Beds  Baths    Size SizeUnit  \
0    12/4/15  710000.00             $     False   1.0    NaN   550.0     sqft   
1   11/30/15       2.15            $M     False   NaN    2.0  1430.0     sqft   
2   11/12/15       5.60            $M     False   2.0    3.5  2040.0     sqft   
3   12/11/15       1.50            $M     False   1.0    1.0  1060.0     sqft   
4    1/15/16  970000.00             $     False   2.0    2.0  1299.0     sqft   

   LotSize LotSizeUnit  BuiltInYear  
0      NaN         NaN       1980.0  
1   2435.0        sqft       1948.0  
2   3920.0        sqft       1976.0  
3      NaN         NaN       1930.0  
4      NaN         NaN       1993.0

> ## `.tail()`: last 5 (default) rows

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tail.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.tail.html)

In [10]:
# TODO
df.tail()

ID                                  Address  Latitude  Longitude  \
995  2124214951    412 Green St APT A, San Francisco, CA  37800040 -122406100   
996  2126960082  355 1st St UNIT 1905, San Francisco, CA  37787029 -122393638   
997  2128308939     33 Santa Cruz Ave, San Francisco, CA  37709136 -122465332   
998  2131957929        1821 Grant Ave, San Francisco, CA  37803760 -122408531   
999  2136213970         1200 Gough St, San Francisco, CA  37784770 -122424100   

    DateOfSale  SalePrice SalePriceUnit IsAStudio  Beds  Baths    Size  \
995    1/15/16   390000.0             $      True   NaN    1.0   264.0   
996   11/20/15   860000.0             $     False   1.0    1.0   691.0   
997   12/10/15   830000.0             $     False   3.0    3.0  1738.0   
998   12/15/15   835000.0             $     False   2.0    2.0  1048.0   
999    1/10/16   825000.0             $     False   1.0    1.0   900.0   

    SizeUnit  LotSize LotSizeUnit  BuiltInYear  
995     sqft      NaN         NaN       2012.0  
996     sqft      NaN         NaN       2004.0  
997     sqft   2299.0        sqft       1976.0  
998     sqft      NaN         NaN       1975.0  
999     sqft      NaN         NaN       1966.0

> ## `.shape`: shape (i.e., number of rows and columns)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.shape.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.shape.html)

In [11]:
df.shape

(1000, 15)

The first value (at index 0) is the number of rows, the second (at index 1), the number of columns:

In [12]:
df.shape[0]

1000

In [ ]:
# TODO

You can also use the idiomatic Python `len` function to get the number of rows:

In [13]:
len(df)

1000

> ## `.dtypes`: column types

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dtypes.html)
- (http://pandas.pydata.org/pandas-docs/stable/basics.html)

In [14]:
df.dtypes

ID                 int64
Address           object
Latitude           int64
Longitude          int64
DateOfSale        object
SalePrice        float64
SalePriceUnit     object
IsAStudio         object
Beds             float64
Baths            float64
Size             float64
SizeUnit          object
LotSize          float64
LotSizeUnit       object
BuiltInYear      float64
dtype: object

> ## `.isnull()` and `.notnull()`: NaN (Not-a-Number)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.isnull.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.isnull.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.isnull.html)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.notnull.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.notnull.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.notnull.html)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sum.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sum.html)

As a data scientist, we will have to decide what to do when encountering missing values (a.k.a, not-a-numbers).  We might decide to drop the row containing it, drop the whole column, or impute it.  Today, let's focus on finding these NaNs.

In [15]:
df.isnull()

ID  Address  Latitude  Longitude  DateOfSale  SalePrice  \
0    False    False     False      False       False      False   
1    False    False     False      False       False      False   
2    False    False     False      False       False      False   
3    False    False     False      False       False      False   
4    False    False     False      False       False      False   
5    False    False     False      False       False      False   
6    False    False     False      False       False      False   
7    False    False     False      False       False      False   
8    False    False     False      False       False      False   
9    False    False     False      False       False      False   
10   False    False     False      False       False      False   
11   False    False     False      False       False      False   
12   False    False     False      False       False      False   
13   False    False     False      False       False      False   
14   False    False     False      False       False      False   
15   False    False     False      False       False      False   
16   False    False     False      False       False      False   
17   False    False     False      False       False      False   
18   False    False     False      False       False      False   
19   False    False     False      False       False      False   
20   False    False     False      False       False      False   
21   False    False     False      False       False      False   
22   False    False     False      False       False      False   
23   False    False     False      False       False      False   
24   False    False     False      False       False      False   
25   False    False     False      False       False      False   
26   False    False     False      False       False      False   
27   False    False     False      False       False      False   
28   False    False     False      False       False      False   
29   False    False     False      False       False      False   
..     ...      ...       ...        ...         ...        ...   
970  False    False     False      False       False      False   
971  False    False     False      False       False      False   
972  False    False     False      False       False      False   
973  False    False     False      False       False      False   
974  False    False     False      False       False      False   
975  False    False     False      False       False      False   
976  False    False     False      False       False      False   
977  False    False     False      False       False      False   
978  False    False     False      False       False      False   
979  False    False     False      False       False      False   
980  False    False     False      False       False      False   
981  False    False     False      False       False      False   
982  False    False     False      False       False      False   
983  False    False     False      False       False      False   
984  False    False     False      False       False      False   
985  False    False     False      False       False      False   
986  False    False     False      False       False      False   
987  False    False     False      False       False      False   
988  False    False     False      False       False      False   
989  False    False     False      False       False      False   
990  False    False     False      False       False      False   
991  False    False     False      False       False      False   
992  False    False     False      False       False      False   
993  False    False     False      False       False      False   
994  False    False     False      False       False      False   
995  False    False     False      False       False      False   
996  False    False     False      False       False      False   
997  False    False     False      False       False   

In return, we get a new `DataFrame` with Boolean values.  `True` if the value is `NaN`, `False` otherwise.

We can also get the count per column:

In [ ]:
df.isnull().sum()

Summing again will return the number of cells in the `DataFrame` with missing values.

In [ ]:
# TODO

Equivalently, we can also use the `.isnull()` function:

In [ ]:
pd.isnull(df)

We also also use `.notnull()`, its complement method:

In [ ]:
df.notnull()

In [ ]:
pd.notnull(df)

> ### `.index` and `.columns`: row and column labels

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.html)

Use the `.index` property to get the label for rows.  For columns, use the `.columns` property.

In [ ]:
df.index

In [ ]:
type(df.index)

In this specific case, rows are just numbered from 0 to 1,000.  Note that, similarly to Python's standard `range` function, this range also excludes the last number.

In [ ]:
df.columns

In [ ]:
type(df.columns)

> ## `[ [] ]` and `[]`: subsetting on columns

Selecting specific columns is performed by using the `[]` operator.

Passing a single integer, or a list of integers, to `[]` will perform a location-based lookup of the columns.

> E.g., columns #5 and #6:

In [ ]:
# TODO

> Let's check that the column subsetting returns a `DataFrame`:

In [ ]:
# TODO

> Note that a `DataFrame` can be made of a single column.  E.g., column #7 only:

In [ ]:
# TODO

If the values passed to `[]` are non-integers, the `DataFrame` will attempt to match them to those in the `columns` property.

> Let's subset the `DataFrame` on columns `SalePrice` and `SalePriceUnit`:

In [ ]:
# TODO

However, you cannot mix integers and non-integers.  E.g.,

In [ ]:
# "df[ ['SalePrice', 6] ]" errors out...  Try it!

Not passing a list always results in a value-based lookup of the column:

In [ ]:
df['Address']

> ## `Series`

(http://pandas.pydata.org/pandas-docs/stable/dsintro.html)

> Let's check the result type:

In [ ]:
# TODO

Columns can also be retrieved using "attribute" access as `DataFrame`s add a property for each column with the names of the properties as the names of the columns.  This won't work however for columns that have spaces or dots in their name.

> Let's check the value of `df`'s `.Address` property:

In [ ]:
# TODO

> Use the `.name` property (not `.columns`, that's for a `DataFrame`) to get the name of the variable stored inside it.

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.name.html)

In [ ]:
# TODO

> To find the zero-based location of a column, use the `.get_loc()` method of the `.columns` property.  E.g.,

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.get_loc.html)

In [ ]:
df.columns.get_loc('Beds')

In [ ]:
df[ [df.columns.get_loc('Beds')] ]

We should get the same output as subsetting a `DataFrame` on `Beds`:

In [ ]:
# TODO

> ## `[]`: slicing on rows

> E.g., on the first five rows:

In [ ]:
# TODO

> ## `.loc[]` and `.iloc[]`: subsetting rows by index label and location

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.loc.html)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.iloc.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.iloc.html)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.set_index.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.set_index.html)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.reset_index.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.reset_index.html)

Until now, the index of the `DataFrame` is a numerical starting from 0 but you can specify which column(s) should be in the index.

> E.g., `ID`:

In [ ]:
df = df.set_index('ID')

In [ ]:
df.index

In [ ]:
df

> E.g., row with index 15063505:

In [ ]:
# TODO

Its name is its value in the index.

> E.g., rows with indices 15063505 and 15064044:

In [ ]:
# TODO

> E.g., rows #1 and #3:

In [ ]:
# TODO

> ## Subsetting rows by Boolean selection (a.k.a., masking)

Rows can also be selected by using Boolean selection, using an array calculated from the result of applying a logical condition on the values in any of the columns.  This allows us to build more complicated selections than those based simply upon index labels or positions.

> E.g., what homes have been built before 1900?

In [ ]:
# TODO

> Let's subset on that `Series`:

In [ ]:
# TODO

Multiple conditions can be put together.

> E.g., subset for `BuiltInYear` below 1900 and `Size` over 1500:

In [ ]:
# TODO

It is possible to subset on columns simultaneously.

> E.g., subset (a `DataFrame`) on `Address` for `BuiltInYear` below 1900 and `Size` over 1500:

In [ ]:
# TODO

> To get a `Series` instead of a `DataFrame`:

In [ ]:
# TODO

## Part B - Wrangling the SF Housing dataset (take 2) with `pandas`

In [ ]:
df = pd.read_csv(os.path.join('..', 'datasets', 'dataset-02-zillow-starter.csv'), index_col = 'ID')

(`pd.read_csv` can load the dataset and set the index column for the `DataFrame` at the same time)

> ### Remove the `Latitude` and `Longitude` columns

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [ ]:
df.drop(['Latitude', 'Longitude'], axis = 1, inplace = True)

In [ ]:
df

> ### `SalePrice`: scale all amount to `$M`

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.unique.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_index.html)

In [ ]:
df.SalePriceUnit.unique()

In [ ]:
df_1 = df[df.SalePriceUnit == '$']
df_1 = df_1.drop('SalePriceUnit', axis = 1)

# Scaling sale price to $M
df_1.SalePrice /= 10 ** 6

df_6 = df[df.SalePriceUnit == '$M']
df_6 = df_6.drop('SalePriceUnit', axis = 1)

In [ ]:
# Concatenate of two DataFrames by rows
df = pd.concat([df_1, df_6])

In [ ]:
# Resort the new DataFrame
df.sort_index(inplace = True)

> ### `IsAStudio`: convert from a Boolean to a binary variable (i.e., 0 or 1)

In [ ]:
# TODO

In [ ]:
df.IsAStudio

> ### `Size`

In [ ]:
df.SizeUnit.unique()

Size is either in square feet or missing.  Almost no work needed except to remove size unit.

In [ ]:
df.drop('SizeUnit', axis = 1, inplace = True)

> ### `LotSize`: scale all values to square feet

In [16]:
df.LotSizeUnit.unique()

array([nan, 'sqft', 'ac'], dtype=object)

Lot sizes are either in square feet or in acres.  Let's convert them all to square feet.

> Group #1: the `na` values:

In [17]:
df_na = df[df.LotSizeUnit.isnull()]
df_na = df_na.drop('LotSizeUnit', axis = 1)

df_na.shape[0]

444

> Group #2: the `sqft` values:

In [20]:
# TODO (use df_sqft)
df_sqft = df[df.LotSizeUnit == 'sqft']
df_sqft.shape[0]

552

> Group #3: the `ac` values:

In [ ]:
# TODO (use df_ac)
df_ac = df[df.LotSizeUnit == 'ac']
df_ac.shape[0]

> Let's scale these `acre` values into `sqft`:

In [ ]:
# (1 acre = 43,560 sqft)

# TODO

Let's now put everything back together...

In [ ]:
df = pd.concat([df_na, df_sqft, df_ac]).sort_index()

In [ ]:
df

> ## `.to_csv`: save the `DataFrame` into a `.csv` file

At the end of each phase (i.e., wrangling) of your data science project, it is a good idea to save your dataset into disk.  Then for the next step, create a new Jupyther notebook and load your updated dataset

In [ ]:
df.to_csv(os.path.join('..', 'datasets', 'dataset-02-zillow.csv'), index_label = 'ID')

## Part C - More advanced topics

### `.groupby()`

(http://pandas.pydata.org/pandas-docs/stable/groupby.html)

> What is the mean price of houses by number of bedrooms?

In [ ]:
df = pd.read_csv()

In [ ]:
df[ ['Beds', 'SalePrice'] ].groupby('Beds').mean()

### `.map()`

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html)

When converting `SalePrice`, `Size`, and `LotSize` into `$M` and sqft, we could also have done the following:

In [ ]:
df = pd.read_csv(os.path.join('..', 'datasets', 'dataset-02-zillow-starter.csv'))

In [ ]:
df.SalePriceUnit.unique()

In [ ]:
df.SalePriceUnit.map({'$': 1. / (10 ** 6), '$M': 1.})

In [ ]:
df.SalePrice *= df.SalePriceUnit.map({'$': 1. / (10 ** 6), '$M': 1.})

In [ ]:
df.SalePrice

In [ ]:
df.drop('SalePriceUnit', axis = 1, inplace = True)

> ### Activity:  Using `.map()`, convert `Size` and `LotSize` to sqft.

In [ ]:
# TODO

### `.to_datetime()`

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)

In [ ]:
df.DateOfSale

So far, the dates stored in the `DataFrame` are just strings.  We cannot easily extract the day, month, year.  Thanksfully, `pandas` provides some facilities to do so.

In [ ]:
pd.to_datetime(df.DateOfSale)

In [ ]:
df.DateOfSale = pd.to_datetime(df.DateOfSale)

### `.apply()`

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html)

In [ ]:
df.DateOfSale.apply(lambda date_of_sale: date_of_sale.year)

In [ ]:
df['YearOfSale'] = df.DateOfSale.apply(lambda date_of_sale: date_of_sale.year)
df['MonthOfSale'] = df.DateOfSale.apply(lambda date_of_sale: date_of_sale.month)
df['DayOfSale'] = df.DateOfSale.apply(lambda date_of_sale: date_of_sale.day)
df['WeekDayOfSale'] = df.DateOfSale.apply(lambda date_of_sale: date_of_sale.weekday_name)

df.drop('DateOfSale', axis = 1, inplace = True)

Now, we have the day, day of the week, month, and year of the sale as features in our dataset.

In [ ]:
df